github/client.go

```diff
@@ -0,0 +1,248 @@
+func WithConcurrency(concurrency int64) ClientOption {
+	return func(options *ClientOptions) {
+		if concurrency >= 1 {
+			options.concurrency = concurrency
+		}
+	}
+}
+
+// WithGraphQLAPIURL sets GitHub's base GraphQL API URL.
+func WithGraphQLAPIURL(graphQLAPIURL string) ClientOption {
+	return func(options *ClientOptions) {
+		options.graphQLAPIURL = graphQLAPIURL
+	}
+}
+
+// WithHeaders adds additional HTTP headers
+func WithHeaders(headers map[string]string) ClientOption {
+	return func(options *ClientOptions) {
+		options.headers = headers
+	}
+}
+
+func paginate[T any, L any](ctx context.Context, c *Client, reqData GraphQLRequest, pageExtractor func(*T) []*L, pageInfoExtractor func(*T) PageInfo) ([]*L, error) {
+	entries := make([]*L, 0, 10)
+	hasNextPage := true
+	endCursor := ""
+	for hasNextPage {
+		if endCursor != "" {
+			reqData.Variables["after"] = endCursor
+		}
+
+		resData, err := doRequest[T](ctx, c, reqData)
+		if err != nil {
+			return entries, errors.Wrap(err, "pagination error")
+		}
+
+		entries = append(entries, pageExtractor(resData)...)
+
+		pageInfo := pageInfoExtractor(resData)
+		if pageInfo.HasNextPage {
+			endCursor = pageInfo.EndCursor
+		} else {
+			hasNextPage = false
+		}
+	}
+
+	return entries, nil
+}
+
+func doRequest[T any](ctx context.Context, c *Client, reqData GraphQLRequest) (*T, error) {
+	req, err := c.createRequestWithBody(ctx, reqData)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := c.doRequestWithConcurrency(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	gqlRes, err := handleGraphQLResponse(res)
+	if err != nil {
+		return nil, err
+	}
+
+	err = handleErrors(gqlRes)
+	if err != nil {
+		return nil, err
+	}
+
+	resData, err := toResponseData[T](gqlRes)
+	if err != nil {
+		return nil, err
+	}
+	return resData, nil
+}
+
+func (c *Client) createRequestWithBody(ctx context.Context, reqData GraphQLRequest) (*http.Request, error) {
+	b, err := json.Marshal(reqData)
+	if err != nil {
+		return nil, errors.Wrap(err, "request marshalling error")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.url, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, errors.Wrap(err, "request error")
+	}
+
+	for k, v := range c.headers {
+		req.Header.Set(k, v)
+	}
+	return req, nil
+}
+
+func (c *Client) doRequestWithConcurrency(ctx context.Context, req *http.Request) (*http.Response, error) {
+	err := c.concurrencySemaphore.Acquire(ctx, int64(1))
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot acquire semaphore")
+	}
+	defer c.concurrencySemaphore.Release(int64(1))
+	res, err := c.http.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "http request call error")
+	}
+	return res, nil
+}
+
+func handleGraphQLResponse(res *http.Response) (*GraphQLResponse, error) {
+	defer res.Body.Close()
+
+	resBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "response body read error")
```

Comment:
nit: We could probably return `ErrorWithStatusCode` here, might be useful for debugging